go-controller/pkg/clustermanager/network_cluster_controller.go

```diff
@@ -781,14 +781,22 @@ func newIPAllocatorForNetwork(netInfo util.NetInfo) (subnet.Allocator, error) {
 
 	subnets := netInfo.Subnets()
 	ipNets := make([]*net.IPNet, 0, len(subnets))
-	excludeSubnets := netInfo.ExcludeSubnets()
+	excludeSubnets := append(netInfo.ExcludeSubnets(), netInfo.InfrastructureSubnets()...)
+
 	for _, subnet := range subnets {
 		ipNets = append(ipNets, subnet.CIDR)
 	}
-	if isLayer2UserDefinedPrimaryNetwork(netInfo) {
+
+	if isLayer2UserDefinedPrimaryNetwork(netInfo) && len(netInfo.InfrastructureSubnets()) == 0 {
```

Comment:
hmm what if defaultgateway and infrastructure subnets are set together? then are we excluding full infrastructure subnet from being candidate for IPAM? and in API level we check default gateway is always subnet or contained within infrastructure subnets right?